prov/efa/src/rxr/rxr_msg.c

```diff
@@ -216,95 +144,126 @@ ssize_t rxr_msg_post_rtm(struct rxr_ep *rxr_ep, struct rxr_tx_entry *tx_entry)
 		 * Have the remote side issue a read to copy the data instead
 		 * to work around this issue.
 		 */
-		if (tx_entry->total_len > max_rtm_data_size || efa_mr_is_hmem(tx_entry->desc[0]))
-			/*
-			 * Read message support
-			 * FI_DELIVERY_COMPLETE implicitly.
-			 */
-			ctrl_type = RXR_LONGREAD_MSGRTM_PKT;
-		else
-			ctrl_type = delivery_complete_requested ? RXR_DC_EAGER_MSGRTM_PKT : RXR_EAGER_MSGRTM_PKT;
+		if (tx_entry->total_len > rxr_env.shm_max_medium_size || efa_mr_is_hmem(tx_entry->desc[0]))
+			return RXR_LONGREAD_MSGRTM_PKT + tagged;
 
-		return rxr_pkt_post_ctrl(rxr_ep, RXR_TX_ENTRY, tx_entry, ctrl_type + tagged, 0, 0);
+		return RXR_EAGER_MSGRTM_PKT + tagged;
 	}
 
-	ret = efa_ep_use_p2p(efa_ep, tx_entry->desc[0]);
-	if (ret < 0)
-		return ret;
-	if (ret == 1 && efa_mr_is_cuda(tx_entry->desc[0])) {
-		return rxr_msg_post_cuda_rtm(rxr_ep, tx_entry);
-	}
+	if (use_p2p && efa_mr_is_cuda(tx_entry->desc[0]))
+		return rxr_msg_select_rtm_for_cuda(rxr_ep, tx_entry);
 
-	/* inter instance message */
-	if (tx_entry->total_len <= max_rtm_data_size) {
-		ctrl_type = (delivery_complete_requested) ?
-			RXR_DC_EAGER_MSGRTM_PKT : RXR_EAGER_MSGRTM_PKT;
-		return rxr_pkt_post_ctrl(rxr_ep, RXR_TX_ENTRY, tx_entry,
-					 ctrl_type + tagged, 0, 0);
-	}
+	/* inter instance message using host/neuron memory */
+	if (tx_entry->fi_flags & FI_INJECT)
+		delivery_complete_requested = false;
+	else
+		delivery_complete_requested = tx_entry->fi_flags & FI_DELIVERY_COMPLETE;
+
+	eager_rtm = (delivery_complete_requested) ? RXR_DC_EAGER_MSGRTM_PKT + tagged
+						  : RXR_EAGER_MSGRTM_PKT + tagged;
+
+	medium_rtm = (delivery_complete_requested) ? RXR_DC_MEDIUM_MSGRTM_PKT + tagged
+						   :  RXR_MEDIUM_MSGRTM_PKT + tagged;
+
+	longcts_rtm = (delivery_complete_requested) ? RXR_DC_LONGCTS_MSGRTM_PKT + tagged
+						    : RXR_LONGCTS_MSGRTM_PKT + tagged;
+
+	longread_rtm = RXR_LONGREAD_MSGRTM_PKT + tagged;
+
+	eager_rtm_max_data_size = rxr_pkt_req_max_data_size(rxr_ep,
+							    tx_entry->addr,
+							    eager_rtm,
+							    tx_entry->fi_flags, 0);
+
+	if (tx_entry->total_len <= eager_rtm_max_data_size)
+		return eager_rtm;
 
 	/*
 	 * Force the LONGREAD protocol for Neuron buffers, regardless of what
 	 * is specified by the user for protocol switch over points.
 	 */
-	if (efa_mr_is_neuron(tx_entry->desc[0])) {
-		/*
-		 * It is possible for the remote endpoint to support RDMA read,
-		 * but not p2p transfers between efa and neuron. That scenario
-		 * will cause a fatal error; if we want to catch this we will
-		 * need to extend the handshake packet to report device p2p
-		 * support.
-		 */
-		ret = rxr_ep_determine_rdma_support(rxr_ep, tx_entry->addr, peer);
-		if (ret < 0)
-			return ret;
+	if (efa_mr_is_neuron(tx_entry->desc[0]))
```

Comment:
Not related to this refactor but i am wondering if we should add a `assert(use_p2p)` for neuron.

Reply:
Yes, though I think we need to discuss whether it is necessary to check `use_p2p` on send path.